redis/src/aio/connection.rs

```diff
@@ -256,76 +258,277 @@ where
 
-/// Represents a `Monitor` connection.
-pub struct Monitor<C = Pin<Box<dyn AsyncStream + Send + Sync>>>(Connection<C>);
+fn stream_dropped_error() -> RedisError {
+    RedisError::from((
+        ErrorKind::ResponseDropped,
+        "Stream was dropped, response cannot be received",
+    ))
+}
 
 impl<C> PubSub<C>
 where
     C: Unpin + AsyncRead + AsyncWrite + Send,
 {
-    fn new(con: Connection<C>) -> Self {
-        Self(con)
+    fn new(con: Connection<C>) -> PubSub<C> {
+        let framed = ValueCodec::default().framed(con.con);
+
+        PubSub {
+            framed,
+            queue: Default::default(),
+        }
+    }
+
+    /// Splits the [`PubSub`] into a [`SplitPubSubSink`] and [`SplitPubSubStream`].
+    pub fn split(self) -> (SplitPubSubSink<C>, SplitPubSubStream<C>) {
+        SplitPubSubSink::new(self.framed)
+    }
+
+    /// Query the connection with a command and retrieves the response.
+    pub async fn query<'a, T>(&'a mut self, cmd: &'a Cmd) -> RedisResult<T>
+    where
+        T: FromRedisValue,
+    {
+        let mut out = Vec::new();
+
+        cmd.write_packed_command(&mut out);
+        self.framed.send(out).await?;
+
+        self.next_response().await
+    }
+
+    async fn next_response<T>(&mut self) -> RedisResult<T>
+    where
+        T: FromRedisValue,
+    {
+        while let Some(item) = self.framed.next().await {
+            match item {
+                Ok(Ok(value)) => match Msg::from_value(&value) {
+                    Some(msg) => self.queue.push_front(msg),
+                    _ => return from_redis_value(&value),
+                },
+                _ => break,
+            }
+        }
+
+        Err(stream_dropped_error())
     }
 
     /// Subscribes to a new channel.
     pub async fn subscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("SUBSCRIBE").arg(channel).query_async(&mut self.0).await
+        self.query(cmd("SUBSCRIBE").arg(channel)).await
     }
 
     /// Subscribes to a new channel with a pattern.
     pub async fn psubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
-            .await
+        self.query(cmd("PSUBSCRIBE").arg(pchannel)).await
     }
 
     /// Unsubscribes from a channel.
     pub async fn unsubscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("UNSUBSCRIBE")
-            .arg(channel)
-            .query_async(&mut self.0)
-            .await
+        self.query(cmd("UNSUBSCRIBE").arg(channel)).await
     }
 
     /// Unsubscribes from a channel with a pattern.
     pub async fn punsubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PUNSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
+        self.query(cmd("PUNSUBSCRIBE").arg(pchannel)).await
+    }
+}
+
+impl Stream for PubSub {
+    type Item = Result<Msg, RedisError>;
```

Comment:
why is it `RedisResult<Msg>` and not `Msg`?
This is a breaking change -  
https://github.com/redis-rs/redis-rs/blob/5e977e93a29622fdd105b9e8776e771c88e575e3/redis/src/aio/connection.rs#L306

Reply:
Should the stream end on errors? Feels odd to consume errors

Reply:
when does the server send errors over a pubsub connection, and does it terminate the connection?
If the server might send errors that aren't responses to requests from the sink, then yes - pass them on. If not, then there's no need for the result.